rnn_speech_recognition/pytorch/common/data/dali/iterator.py

```diff
@@ -0,0 +1,109 @@
+import numpy as np
+from common.helpers import print_once
+from common.text import _clean_text, punctuation_map
+
+
+def normalize_string(s, charset, punct_map):
+    """Normalizes string.
+
+    Example:
+        'call me at 8:00 pm!' -> 'call me at eight zero pm'
+    """
+    charset = set(charset)
+    try:
+        text = _clean_text(s, ["english_cleaners"], punct_map).strip()
+        return ''.join([tok for tok in text if all(t in charset for t in tok)])
+    except:
+        print(f"WARNING: Normalizing failed: {s}")
+        return None
+
+
+class DaliRnntIterator(object):
+    """
+    Returns batches of data for RNN-T training:
+    preprocessed_signal, preprocessed_signal_length, transcript, transcript_length
+
+    This iterator is not meant to be the entry point to Dali processing pipeline.
+    Use DataLoader instead.
+    """
+
+    def __init__(self, dali_pipelines, transcripts, tokenizer, batch_size, shard_size, pipeline_type, normalize_transcripts=False):
+        self.normalize_transcripts = normalize_transcripts
+        self.tokenizer = tokenizer
+        self.batch_size = batch_size
+        from nvidia.dali.plugin.pytorch import DALIGenericIterator
+        from nvidia.dali.plugin.base_iterator import LastBatchPolicy
+
+        # in train pipeline shard_size is set to divisable by batch_size, so PARTIAL policy is safe
+        if pipeline_type == 'val':
+            self.dali_it = DALIGenericIterator(
+                dali_pipelines, ["audio", "label", "audio_shape"], reader_name="Reader",
+                dynamic_shape=True, auto_reset=True,
+                last_batch_policy=LastBatchPolicy.PARTIAL)
+        else:
+            self.dali_it = DALIGenericIterator(
+                dali_pipelines, ["audio", "label", "audio_shape"], size=shard_size,
+                dynamic_shape=True, auto_reset=True, last_batch_padded=True,
+                last_batch_policy=LastBatchPolicy.PARTIAL)
+
+        self.tokenize(transcripts)
+
+    def tokenize(self, transcripts):
+        transcripts = [transcripts[i] for i in range(len(transcripts))]
+        if self.normalize_transcripts:
+            transcripts = [
+                normalize_string(
+                    t,
+                    self.tokenizer.charset,
+                    punctuation_map(self.tokenizer.charset)
+                ) for t in transcripts
+            ]
+        transcripts = [self.tokenizer.tokenize(t) for t in transcripts]
+        transcripts = [torch.tensor(t) for t in transcripts]
+        self.tr = np.array(transcripts, dtype=object)
+        self.t_sizes = torch.tensor([len(t) for t in transcripts], dtype=torch.int32)
+
+    def _gen_transcripts(self, labels, normalize_transcripts: bool = True):
+        """
+        Generate transcripts in format expected by NN
+        """
+        ids = labels.flatten().numpy()
+        transcripts = self.tr[ids]
+        transcripts = torch.nn.utils.rnn.pad_sequence(transcripts, batch_first=True)
```

Comment:
this is a very subtle point - but can be confusing for some. the transcript sequences are padded with values of 0. Note that the blank-symbol for the RNN-T loss has id 1023 and not 0. Fortunately, the padding-value of zero works here, because the sentencepiece code assigns the id 0 to the `<UNK>` symbol which does NOT exist in the dataset. If the sentencepiece code had assigned a different symbol to id 0, this could have led to problems. Might be useful to explicitly specify `fill_value` here and have a comment so that its not confused with the RNN-T blank.

Reply:
Let me double-check if filling with blank doesn't break anything.
It should not matter, because the RNN-T loss implementation respects sequence lengths, but using blanks here makes more sense I think.

Reply:
I talked to Adrian, and he suggests leaving padding with zeroes.
The argument against changing the values is that the user could assume the default value somewhere else, and setting an arbitrary value could complexify it. As the sentencepieces library doesn't support disabling the UNK symbol ([sources](https://github.com/google/sentencepiece/blob/master/src/spm_train_main.cc#L119-L120)), we can make use of it.
RNN-T doesn't read padded areas, as there are no convolutions, and LSTMs and RNN-T loss skip out-of-range data.

Are you ok to keep it unchanged?

Reply:
I added a comment (https://github.com/mlcommons/training/pull/430/commits/ea7a2f30680ab7888f7987b0085d18cf026efba1) clarifying the zero-padding. Please check if it satisfies you.

Reply:
I am okay with keeping this unchanged. Also agree that the padding-value should not matter because the RNN-T loss implementation takes into account sequence lengths. So the comment added is sufficient. Thank you.